src/audio/buffer.c

```diff
@@ -51,23 +52,25 @@ struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t flags, u
 		return NULL;
 	}
 
-	buffer->stream.addr = rballoc_align(0, caps, size, align);
-	if (!buffer->stream.addr) {
+	stream_addr = rballoc_align(0, caps, size, align);
+	if (!stream_addr) {
 		rfree(buffer);
 		tr_err(&buffer_tr, "buffer_alloc(): could not alloc size = %u bytes of type = %u",
 		       size, caps);
 		return NULL;
 	}
 
+	/* From here no more uncached access to the buffer object, except its list headers */
+	buffer_c = buffer_acquire(buffer);
+	buffer_c->stream.addr = stream_addr;
+	buffer_init(buffer_c, size, caps);
+
 	buffer->stream.underrun_permitted = !!(flags & SOF_BUF_UNDERRUN_PERMITTED);
 	buffer->stream.overrun_permitted = !!(flags & SOF_BUF_OVERRUN_PERMITTED);
 
 	list_init(&buffer->source_list);
 	list_init(&buffer->sink_list);
```

Comment:
these buffer allocation and initialisation code is extremely sensitive. You say, that you don't want to mix cached and uncached access, but here you keep uncached access in 4 assignments above. Further, this change should actually corrupt buffer memory: in your new line 64 you obtain a cached address, then you modify the object via that address, i.e. it gets read into cache and the cache copy gets modified. Then the above 4 lines perform modifications on uncached memory, now cache is outdated. Then in line 86 below cache is written back, overwriting your modifications to uncached memory.